src/crypto/tls/tls_ech.go

```diff
@@ -0,0 +1,684 @@
+// Copyright 2020 Cloudflare, Inc. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/cisco/go-hpke"
+	"github.com/cisco/go-tls-syntax" // TODO(cjpatton): Remove this dep
+	"github.com/cjpatton/ech-mint"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+const (
+	// Constants for HPKE operations
+	echLabelAcceptConfirm         = "ech accept confirmation"
+	echHpkeInfoInnerDigest        = "tls13 ech inner digest"
+	echHpkeInfoConfigId           = "tls13 ech config id"
+	echHpkeInfoSetup              = "tls13 ech"
+	echHpkeInfoSetupHrr           = "tls13 ech hrr"
+	echHpkeHrrKeyExportechContext = "tls13 ech hrr key"
+	echHpkeHrrKeyId               = "hrr key"
+	echHpkeHrrKeyLen              = 16
+)
+
+// When providing covertext for the ECH extension, this configuration is
+// used to generate the dummy ECH extension.
+var echDummyConfig = []byte{
+	255, 8, 0, 79, 0, 19, 99, 108, 111, 117, 100, 102, 108, 97, 114, 101, 45,
+	101, 115, 110, 105, 46, 99, 111, 109, 0, 32, 28, 182, 205, 118, 103, 238,
+	194, 134, 248, 202, 133, 203, 15, 98, 30, 77, 28, 28, 44, 47, 254, 187, 87,
+	66, 50, 1, 183, 126, 71, 75, 85, 73, 0, 32, 0, 16, 0, 1, 0, 1, 0, 1, 0, 3,
+	0, 2, 0, 1, 0, 2, 0, 3, 0, 0, 0, 0,
+}
+
+// ECHProvider specifies the interface of an ECH service provider that decrypts
+// ClientHelloInner on behalf of the client-facing server and defines the set of
+// acceptable ECH configurations.
+type ECHProvider interface {
+	// SetupServerContext attempts to construct the decryption context
+	// corresponding to handle, which encodes the ciphersuite and encapsulated
+	// key and indicates the ECH configuration to use for decapsulation.
+	// version specifies the version of ECH indicated by the client. In case of
+	// HelloRetryRequest, hrrPsk specifies the pre-shared key that binds the
+	// first ClientHello to the second.
+	SetupServerContext(handle, hrrPsk []byte, version uint16) (res ECHProviderResult, err error)
```

Comment:
`version` should be removed here as it is is not advertised through the TLS wireformat. Instead, the application developer is expected to ensure that the `SetupServerContext` implementation uses the same version as included with the `ECHConfig` that is matched by the `handle` parameter.

Other suggestions:
- Clarify what `handle` refers to in the documentation, it contains the `cipher_suite`, `config_id`, and `enc` fields of `ClientECH`. Alternatively, just pass those fields explicitly.
- Clarify what the returned "decryption context" is. I hope it is not an implementation detail of go-hpke, and instead something like the "HPKE KeySchedule output" or "HPKE KeySchedule context" (https://tools.ietf.org/html/draft-irtf-cfrg-hpke-05#page-13).

`SetupServerContext(cipher_suite uint16, config_id, enc, hrrPsk []byte)`

Reply:
> `version` should be removed here as it is is not advertised through the TLS wireformat.

It is, in fact. Since https://github.com/tlswg/draft-ietf-tls-esni/pull/303, `ECHConfig.version` is locked to the extension codepoint.

>     * Clarify what `handle` refers to in the documentation, it contains the `cipher_suite`, `config_id`, and `enc` fields of `ClientECH`. Alternatively, just pass those fields explicitly.

I will clarify the meaning of `handle` in the doc. The alternative is risky because future versions of ECH can change the `cipher_suite`, `config_id`, and `enc` fields. (The risk is probably low, but I would rather leave the details of the message format opaque in this API.)

>     * Clarify what the returned "decryption context" is. I hope it is not an implementation detail of go-hpke, and instead something like the "HPKE KeySchedule output" or "HPKE KeySchedule context" (https://tools.ietf.org/html/draft-irtf-cfrg-hpke-05#page-13).7654

Will do. It's not an implementation detail of go-hpke ... it's the HPKE context discussed in the spec.

Reply:
Oh.... I only expected the version in ECHConfig which is not sent over the network. I did not realize that the extension codepoint will be changing with every draft version. That is a significant change which did not seem to receive a lot of comments from others in https://github.com/tlswg/draft-ietf-tls-esni/issues/297.

Since the supported extension version is fixed to the crypto/tls implementation, I think it would be fine to drop the version number. The application developer will have to carefully manage the Go version while working with this experimental feature.

If you want to keep the version number for whatever reason, I won't oppose that. It'll most likely be dropped with the final version anyway.

Reply:
Done.

Reply:
I'm strongly in favor of keeping the version number. Dropping it could become a problem if/when we have to maintain multiple versions simultaneously. In particular, the ECH provider may need to know which version was offered in order to respond correctly. I agree this won't be as problem once we reach RFC, but I suspect there will be revisions to the standard before we get there.